convert2rhel/applock.py

```diff
@@ -0,0 +1,178 @@
+                os.link(fileh.name, self._pidfile)
+            # In Python 3 this could be changed to catch FileExistsError.
+            except OSError as exc:
+                if exc.errno == errno.EEXIST:
+                    return False
+                raise exc
+        loggerinst.debug("%s." % self)
+        return True
+
+    @property
+    def is_locked(self):
+        """Test whether this object is locked."""
+        return self._locked
+
+    @staticmethod
+    def _pid_exists(pid):
+        """Test whether a particular process exists."""
+        try:
+            os.kill(pid, 0)
+        except OSError:
+            return False
+        return True
+
+    def trylock(self, loop_count=0):
+        """Try to get a lock on this application. If successful,
+        the application will be locked; the lock should be released
+        with unlock().
+
+        If the file has unexpected contents, for safety we treat the
+        application as locked, since it is probably the result of
+        manual meddling, intentional or otherwise.
+
+        Note that nothing prevents you from calling trylock()
+        multiple times; all calls after the first will fail as if
+        another process holds the lock.
+
+        :param loop_count: used internally to limit the number of
+                           recursive calls to this method
+        :raises ApplicationLockedError: the application is locked
+        """
+        if loop_count > self._max_loop_count:
+            raise ApplicationLockedError("Cannot lock %s" % self._pidfile)
+        if self._try_create():
+            self._locked = True
+            return
+
+        with io.open(self._pidfile, "rt") as fileh:
+            file_contents = fileh.read()
+            try:
+                pid = int(file_contents.rstrip())
+            except ValueError:
+                raise ApplicationLockedError("Lock file %s is corrupt" % self._name)
+
+        if self._pid_exists(pid):
+            raise ApplicationLockedError("%s locked by process %d" % (self._name, pid))
+        else:
+            #
+            # The lock file was created by a process that has exited;
+            # remove it and try again.
+            #
+            loggerinst.info("Reaping lock held by process %d." % pid)
+            try:
+                os.unlink(self._pidfile)
+            except OSError:
+                loggerinst.warning("Couldn't unlink %s." % self.pidfile)
+            self.trylock(loop_count + 1)
+
+    def unlock(self):
+        """Release the lock on this application."""
+        if self._locked == False:
+            return
+        try:
+            os.unlink(self._pidfile)
+        except OSError:
+            loggerinst.warning("Couldn't unlink %s." % self._pidfile)
+        loggerinst.debug("%s." % self)
+        self._locked = False
+
+    def __enter__(self):
+        self.trylock()
+        return self
+
+    def __exit__(self, _type, _value, _tb):
+        self.unlock()
```

Comment:
I had no strong feelings about how things should behave if you try and grab a second lock, and I actually had a comment noting that if you tried to lock a second time it would fail--I removed it because I wasn't sure I wanted people to count on it, and I didn't write a test.

Since you feel strongly about it, I can do the relocking modification and convert self.locked to a refcount.

Reply:
Joe and I talked about this on a video chat.  I said that I wasn't married to refcounting for this code; my real concern was making sure that the API was consistent and we have a single source of truth.  If we use the lock file for locking operations but use the `self._locked` attribute which tells us whether this particular ApplicationLock instance was used to lock a resource for everything else, then we run into cases where the API can be used to get us into an unexpected state.

Given that we don't need refcounting of the lock for anything that we're doing with it and that it would be more complicated to implement that, we agreed that removing `self._locked` might be the better way to make the API consistent.  When the `is_locked` property is consulted, we can check whether the lock file exists and whether it is something that we (the application, no longer the specific ApplicationLock instance) owns.  That way the file becomes the single source of truth.

Reply:
Recorded the cleanup for this in this jira ticket: https://issues.redhat.com/browse/RHELC-1124